crates/datafusion-optimizer-rules/src/logical_plan/duckdb/aggregate_pushdown.rs

```diff
@@ -0,0 +1,618 @@
+                }
+                _ => return Ok(None),
+            }
+        }
+
+        // Scan its children to ensure that there is a unary chain to an accelerated
+        // DuckDB provider
+        let mut found = false;
+
+        let _ = plan.apply(|p| match p {
+            LogicalPlan::TableScan(table_scan) if Self::is_duckdb_provider(table_scan)? => {
+                found = true;
+                Ok(TreeNodeRecursion::Stop)
+            }
+            other if other.inputs().len() > 1 => Ok(TreeNodeRecursion::Stop),
+            _ => Ok(TreeNodeRecursion::Continue),
+        })?;
+
+        if found {
+            Ok(Some(LogicalPlan::Extension(Extension {
+                node: DuckDBAggregatePushdownNode::new(plan.clone()),
+            })))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Try to find a unary path to a marker node from the current node, then swap it
+    fn try_percolate_marker_node(plan: &LogicalPlan) -> Result<Option<LogicalPlan>> {
+        let with_erased_marker = plan.clone().transform_down(|p| {
+            if p.inputs().len() > 1 || matches!(p, LogicalPlan::Analyze(_)) {
+                return Ok(Transformed::new(p, false, TreeNodeRecursion::Stop));
+            }
+
+            let LogicalPlan::Extension(ref ext) = p else {
+                return Ok(Transformed::no(p));
+            };
+
+            let Some(marker) = ext
+                .node
+                .as_any()
+                .downcast_ref::<DuckDBAggregatePushdownNode>()
+            else {
+                return Ok(Transformed::no(p));
+            };
+
+            Ok(Transformed::new(
+                marker.input_plan.clone(),
+                true,
+                TreeNodeRecursion::Jump,
+            ))
+        })?;
+
+        if with_erased_marker.transformed {
+            Ok(Some(LogicalPlan::Extension(Extension {
+                node: DuckDBAggregatePushdownNode::new(with_erased_marker.data),
+            })))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl OptimizerRule for DuckDBAggregateLogicalPushdown {
+    fn name(&self) -> &'static str {
+        "DuckDBAggregatePushdownOptimizerRule"
+    }
+
+    // This rule does its own recursion
+    fn apply_order(&self) -> Option<ApplyOrder> {
+        None
+    }
+
+    fn rewrite(
+        &self,
+        plan: LogicalPlan,
+        _config: &dyn OptimizerConfig,
+    ) -> Result<Transformed<LogicalPlan>, DataFusionError> {
+        // Mark all eligible nodes for DuckDB agg pushdown
+        let maybe_marked_agg = plan.transform_down(|p| {
+            match &p {
+                // Skip already marked subtrees
+                LogicalPlan::Extension(ext)
+                    if concrete!(ext.node, DuckDBAggregatePushdownNode).is_some() =>
+                {
+                    return Ok(Transformed::new(p, false, TreeNodeRecursion::Jump));
+                }
+                _ => { /* no-op */ }
+            }
```

Comment:
single match arms are usually better represented with a single `if`:

```suggestion
            if let LogicalPlan::Extension(ext) = &p && concrete!(ext.node, DuckDBAggregatePushdownNode).is_some() {
                    return Ok(Transformed::new(p, false, TreeNodeRecursion::Jump));
            }
```

Reply:
TIL you can use destructured bindings after `&&`. In Scala the only place that works is match statements, so it's taking me a bit to unlearn the habit 😄